feat: improve readability and add documentation (#185)
influxdata/giraffe: giraffe/README.md

   - ticks: generation, formatting and labeling, font, and color
   - legend (tooltip): labeling and styling

   For details on all configuration properties, skip to [config](#Config) documentation.

   Here is an example of building the config object while skipping optional properties:

  <pre>
  // Example table and layer

  const table = newTable(5)
    .addColumn('_time', 'time', [1589838401244, 1589838461244, 1589838521244, 1589838581244, 1589838641244])
    .addColumn('_value', 'number', [2.58, 7.11, 4.79, 8.89, 2.23])

  const lineLayer = {
    type: "line",
    x: "_time",
    y: "_value",
  }

  const config = {
    table: table,
    layers: [lineLayer],
  }
  </pre>

3. Render your component by passing the `confg` object as the config prop to the `<Plot>` component. Be sure that the parent component around `<Plot>` has both a height and a width measured in positive values. If either is not a positive value, the graph will not be visible.

   For example, to make a `<Plot>` that adjusts to screen height and width, in your React rendering code return this element:

  <pre>
  // return this element in your React rendering code:

>   &#60;div
>     style={{
>       width: "calc(70vw - 20px)",
>       height: "calc(70vh - 20px)",
>       margin: "40px",
>     }}
>   &#62;
>     &#60;Plot config={config} /&#62;
>   &#60;/div&#62;

Reviewer: ```suggestion
  <div
    style={{
      width: "calc(70vw - 20px)",
      height: "calc(70vh - 20px)",
      margin: "40px",
    }}
  >
    <Plot config={config} />
  </div>
```

Author: We can't use actual divs inside a `<pre>` if we want to show the literal "`<div>`" as text. I chose `<pre>` over triple \` marks because I wanted to indent the whole code block, which `<pre>` allows me to do while \` marks do not.

,,0,2020-03-25T20:58:15.731129Z,2020-04-24T20:58:15.731129Z,2020-04-03T18:31:33.95Z,29.9,value,temperature,index.html,browser
,,0,2020-03-25T20:58:15.731129Z,2020-04-24T20:58:15.731129Z,2020-04-03T18:55:23.863Z,28.7,value,temperature,index.html,browser
,,0,2020-03-25T20:58:15.731129Z,2020-04-24T20:58:15.731129Z,2020-04-03T19:50:52.357Z,15,value,temperature,index.html,browser
,,0,2020-03-25T20:58:15.731129Z,2020-04-24T20:58:15.731129Z,2020-04-03T19:53:37.198Z,24.8,value,temperature,index.html,browser
,,0,2020-03-25T20:58:15.731129Z,2020-04-24T20:58:15.731129Z,2020-04-03T19:53:53.033Z,23,value,temperature,index.html,browser
,,0,2020-03-25T20:58:15.731129Z,2020-04-24T20:58:15.731129Z,2020-04-03T20:19:21.88Z,20.1,value,temperature,index.html,browser
,,0,2020-03-25T20:58:15.731129Z,2020-04-24T20:58:15.731129Z,2020-04-10T22:20:40.776Z,28.7,value,temperature,index.html,browser
`

  const dataFromFlux = fromFlux(fluxResultCSV)

  const lineLayer = {
    type: "line",
    x: "_time",
    y: "_value",
  }

  const config = {
    table: dataFromFlux.table,
    layers: [lineLayer],
  }

  // ...

  // return this element in your React rendering code:

>   &#60;div
>     style={{
>       width: "calc(70vw - 20px)",
>       height: "calc(70vh - 20px)",
>       margin: "40px",
>     }}
>   &#62;
>     &#60;Plot config={config} /&#62;
>   &#60;/div&#62;

Reviewer: ```suggestion
  <div
    style={{
      width: "calc(70vw - 20px)",
      height: "calc(70vh - 20px)",
      margin: "40px",
    }}
  >
    <Plot config={config} />
  </div>
```

Author: We can't use actual divs inside a `<pre>` if we want to show the literal "`<div>`" as text. I chose `<pre>` over triple \` marks because I wanted to indent the whole code block, which `<pre>` allows me to do while \` marks do not.

  </pre>

## Config

`<Plot>` requires a `config` prop which is an object with properties that serve three purposes:

- customizing the [appearance](#appearance-properties) of `<Plot>`
- specifying the [data](#data-properties) to be visualized
- defining the columns and appearance of the [legend (tooltip)](#legend-tooltip-properties).

<pre>
  const config = {
    <b><i>property</i></b>: <b><i>value</i></b>,
    <b><i>...</i></b>
  }
</pre>

### Appearance properties

> - **width**: _number. Optional._ The width in _CSS px_ of the Plot. Includes the space to the left and below the axes surrounding the ticks, tick labels, and axis labels. When not specified, the width of the parent element to `<Plot>` will determine.

Reviewer: ```suggestion
- **width**: _number. Optional._ The width in _CSS px_ of the Plot. Includes the space to the left the axes surrounding the ticks, tick labels, and axis labels. When not specified, the width of the `<Plot>`'s parent element  will used to determine the width.
```

Author: 🎉

> - **height**: _number. Optional._ The height in _CSS px_ of the Plot. Includes the space to the left and below the axes surrounding the ticks, tick labels, and axis labels. When not specified, the height of the parent element to `<Plot>` will determine.

Reviewer: ```suggestion
- **height**: _number. Optional._ The height in _CSS px_ of the Plot. Includes the space to the below the axes surrounding the ticks, tick labels, and axis labels. When not specified, the height of `<Plot>`'s parent element will determine the height.
```

Author: 🎉

- **gridOpacity**: _number. Optional. Recommendation: do not include. Defaults to 1 when excluded._ A value between 0 and 1 for the [_CanvasRenderingContext2D globalAlpha_](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalAlpha) of the grid lines. Applies to the inner horizontal and vertical rule lines. Excludes the axes and the border around the graph.

- **showAxes**: _boolean. Optional. Recommendation: do not include. Defaults to true when excluded._ Indicates whether Plot axes should be visible. Applies to both x-axis and y-axis simultaneously.

- **axisColor**: _string. Optional._ The _CSS color value_ of the axes and the border around the graph. Excludes the inner horizontal and vertical rule lines.

- **axisOpacity**: _number. Optional. Recommendation: do not include. Defaults to 1 when excluded._ A value between 0 and 1 for the [_CanvasRenderingContext2D globalAlpha_](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalAlpha) of the axes and the border around the graph. Excludes the inner horizontal and vertical rule lines.

> - **xTicks**: _array[number, ...]. Optional._ An array of values representing tick marks on the x-axis. Actual data values and axis scaling may cause Plot to not render all of the given ticks, or Plot rendering may extend beyond all of the rendered ticks.

Reviewer: Should we add a little blurb about how the ticks are calculated if none are passed in? Or what the default is?

Author: Yes. 🎉

> - **yTicks**: _array[number, ...]. Optional._ An array of values representing tick marks on the y-axis. Actual data values and axis scaling may cause Plot to not render all of the given ticks, or Plot rendering may extend beyond all of the rendered ticks.

Reviewer: Same as above?

Author: Yup. 🎉

      _value: (num) => num.toFixed(2),                 // values fixed to 2 decimal places
    },
  
    // ...
  }
  </pre>

- **xAxisLabel**: _string. Optional._ Uses **tickFont**, **tickFontColor**. Name to display below the x-axis.

- **yAxisLabel**: _string. Optional._ Uses **tickFont**, **tickFontColor**. Name to display to the left of the y-axis.

### Data properties

- **table**: _Object. **Required**._ A data object produced by [Giraffe's utility functions](#utility-functions). Houses the data and getters and setters for that data, for the `<Plot>`.

  - the `table` property of the return value from the `fromFlux` utility function.
  - the return value from the `fromRows` utility function.
  - the return value from the `newTable` utility function.

- **layers**: _array[Object, ...]. **Required**._ An array of [LayerConfig objects](#layerconfig). These objects are customizations specific to a type of `<Plot>`. Currently, Giraffe supports only one pre-defined `<Plot>` type with any number of "custom layers". Custom layers are not pre-defined in Giraffe, but created through a callback render function in the configuration.

> - **xScale**: _"linear" | "log". Optional._ Sets the scaling of the x-axis by selecting the correct scaling function to be used by Layers. Linear scaling means the same distance between ticks represent the same increase in value. Logarithmic (log) scaling means the same distance between ticks can represent an exponential increase in value, used as a way to display data with a very wide range of values in a compact space.

Reviewer: This is a little confusing. Does the `log` scaling mean `exponential` increases in value on Giraffe or `logarithmic`?

Author: I rewrote this part to be clearer.

> - **yScale**: _"linear" | "log". Optional._ Sets the scaling of the y-axis by selecting the correct scaling function to be used by Layers. Linear scaling means the same distance between ticks represent the same increase in value. Logarithmic (log) scaling means the same distance between ticks can represent an exponential increase in value, used as a way to display data with a very wide range of values in a compact space.

Reviewer: same as above

Author: 🎉

> - **xDomain**: _array[min, max]. Optional._ The x domain of the plot can be explicitly set with numbers denoting a _min_ and _max_. If this option is passed, then `<Plot>` is operating in a "controlled" mode, where it always uses the passed x domain. Any brush interaction with the `<Plot>` that should change the x domain will call the `onSetXDomain` option when the component is in controlled mode. Double clicking the plot will call `onResetXDomain`. If the `xDomain` option is not passed, then the component is "uncontrolled". It will compute, set, and reset the `xDomain` automatically.

Reviewer: ```suggestion
- **xDomain**: _array[min, max]. Optional._ The x domain of the plot can be explicitly set with numbers denoting a _min_ and _max_. If this option is passed, then `<Plot>` is operating in a "controlled" mode, where it always uses the passed x domain to set the minimum and maximum value of the x axis. Any brush interaction with the `<Plot>` that should change the x domain will call the `onSetXDomain` option when the component is in controlled mode. Double clicking the plot will call `onResetXDomain`. If the `xDomain` option is not passed, then the component is "uncontrolled". It will compute, set, and reset the `xDomain` automatically.
```

Reviewer: Also, should we say something about the fact that if a user passes an `xDomain`, it is required that the value have both a `min` and `max` value? So that way users don't input: `[null, 100]`, `[0, null]`, `[0]`, etc....

Author: Good call 🎉

- **onSetXDomain**: _function(array[min, max]). Optional._ See above regarding **xDomain**.

- **onResetXDomain**: _function(). Optional._ See above regarding **xDomain**.

> - **yDomain**: _array[min, max]. Optional._ The y domain of the plot can be explicitly set with numbers denoting a _min_ and _max_. If this option is passed, then `<Plot>` is operating in a "controlled" mode, where it always uses the passed y domain. Any brush interaction with the `<Plot>` that should change the y domain will call the `onSetYDomain` option when the component is in controlled mode. Double clicking the plot will call `onResetYDomain`. If the `yDomain` option is not passed, then the component is "uncontrolled". It will compute, set, and reset the `yDomain` automatically.

Reviewer: ```suggestion
- **yDomain**: _array[min, max]. Optional._ The y domain of the plot can be explicitly set with numbers denoting a _min_ and _max_. If this option is passed, then `<Plot>` is operating in a "controlled" mode, where it always uses the passed y domain to set the minimum and maximum value of the y axis. Any brush interaction with the `<Plot>` that should change the y domain will call the `onSetYDomain` option when the component is in controlled mode. Double clicking the plot will call `onResetYDomain`. If the `yDomain` option is not passed, then the component is "uncontrolled". It will compute, set, and reset the `yDomain` automatically.
```

Reviewer: Same note as above

Author: Good call 🎉

- **legendColumns**: _array[string, ...]. Optional._ When included, this array will determine which column key names that should be included in the legend (tooltip). If this option is included as an empty array, the legend will be empty.

## Utility Functions

Giraffe comes with utility functions.

- **fromFlux**: _function(string)._ Takes a Flux CSV, converts it, and returns a `Table` used in the [table property](#data-properties) of the config.

- **fromRows**: _function([Object, ...], Object)._ The first argument is an array of objects, each representing a row of data. The optional second argument describes the schema for the data. Returns a `Table` used in the [table property](#data-properties) of the config.

- **newTable**: _function(number)._ The argument is a length for a newly created `Table` with no initial data that allows only columns equal to that length to be added. Returns the created `Table`.

## LayerConfig

- **LineLayerConfig**: _Object_. Maximum one per `<Plot>`. Properties are:

>   - **type**: _"line". **Required**_. Specifies that this LayerConfig and `<Plot>` is a line graph.

Reviewer: Isn't the type a `string` that can be an enum value of `line`?

Author: It's very hard to be clear in docs without writing real code. This is pseudo-code to say that the property name is "type" and the value is literally called "line". Not aiming to bring TypeScript semantics into the docs.

Reviewer: Totally. I was still waking up when I was reviewing this and didn't connect that this was a subset of `LineLayerConfig` 🤦

  - **x**: _string. **Required**_. The column key name of the column that should be visualized on the x-axis.

  - **y**: _string. **Required**_. The column key name of the column that should be visualized on the y-axis.

  - **fill**: _array[string, ...]. Optional._ An array of column key names of column filters that should be visualized. If this option is not included, the data in the graph will be interpreted as belonging to a single column.

  - **position**: _"overlaid" | "stacked". Optional._ Indicates whether the line graph's lines have no bearing on other lines (overlaid), or the lines are cumulatives of every line below it, ie [stacked](https://help.infragistics.com/Help/Doc/Silverlight/2011.1/CLR4.0/html/xamWebChart_Stacked_Line_Chart.html).

  - **hoverDimension**: _"x" | "y" | "xy". Optional. Defaults to "x" when not included._ Indicates whether the legend (tooltip) should display all data points along the "x" axis or "y" axis during mouse hover. When "xy" is chosen, the legend will display for a single data point nearest the mouse rather than for all data points along an entire axis.

>   - **maxTooltipRows**: _number. Optional._ The maxium number of data rows to display in the legend (tooltip). Subject to screen size limitations. Scrolling not implemented.

Reviewer: Is there a default value for this? What happens when no number is input

Author: Yes, you're right, this is important enough to mention the default. 🎉

>   - **interpolation**: _string. Optional._ The style of the path between two data points on the same line. For example, "linear" is a straight path between two data points.

Reviewer: Is there a list of `interpolation` values somewhere? It might be useful to include a reference to that here

Author: Good call, we should definitely be thorough and include references. But the downside is that there is only 1 good reference which is d3. We are admitting we rely on d3 right in the docs if we link to d3.

Oh well, I guess we have to do that for now until we stop relying on d3.

  - **lineWidth**: _number. Optional._ The [CanvasRenderingContext2D lineWidth](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineWidth) of each graph line.

  - **colors**: _array[string, ...]. Optional._ An array of _CSS color values_ used as a gradient to give multiple lines in the graph different colors based on the **fill** columns.

>   - **shadeBelow**: _boolean. Optional._ Uses **colors**. Indicates whether the area below each line should be shaded.

Reviewer: defaults to?

Author: Not important enough to mention.
